File: typescript/src/runtime/payment.ts
/**
 * Pluggable payment handler hook for the SDK transport.
 *
 * When the API returns `402 Payment Required`, the transport calls the
 * configured `PaymentHandler` to produce the extra headers (typically
 * `X-Payment`) to attach to the retry. This keeps the SDK free of any
 * chain-specific signing logic, and lets callers plug in a real x402
 * implementation such as `@acedatacloud/x402-client`.
 */

/** Payment requirement as returned by the server in a 402 response. */
export interface PaymentRequirement {
  scheme: string;
  network: string;
  maxAmountRequired: string;
  maxTimeoutSeconds?: number;
  resource?: string;
  description?: string;
  payTo: string;
  asset: string;
  extra?: Record<string, unknown>;
}

/** Shape of a 402 response body. */
export interface PaymentRequiredBody {
  x402Version?: number;
  accepts: PaymentRequirement[];
  error?: string;
}

/** Context passed to a payment handler when a 402 is observed. */
export interface PaymentHandlerContext {
  url: string;
  method: string;
  body?: unknown;
  accepts: PaymentRequirement[];
}

/** Result a payment handler must return. */
export interface PaymentHandlerResult {
  /** Extra headers to attach to the retry (must include `X-Payment`). */
  headers: Record<string, string>;
}

/** A callable that signs/settles a payment and returns retry headers. */
export type PaymentHandler = (
  ctx: PaymentHandlerContext
) => Promise<PaymentHandlerResult> | PaymentHandlerResult;

File: python/src/acedatacloud/_runtime/payment.py
"""Pluggable payment handler hook for the SDK transport.

When the API returns ``402 Payment Required``, the transport calls the
configured payment handler to obtain the extra headers (typically
``PAYMENT-SIGNATURE``) to attach to the retry. This keeps the SDK free of any
chain-specific signing logic, and lets callers plug in an x402
implementation when one becomes available.
"""

from __future__ import annotations

import base64
import json
from collections.abc import Awaitable, Mapping
from typing import Any, Callable, TypedDict, Union


class PaymentRequirement(TypedDict, total=False):
    scheme: str
    network: str
    amount: str
    maxAmountRequired: str
    maxTimeoutSeconds: int
    resource: str
    description: str
    payTo: str
    asset: str
    extra: dict[str, Any]


class PaymentRequiredBody(TypedDict, total=False):
    x402Version: int
    accepts: list[PaymentRequirement]
    error: str


class PaymentHandlerContext(TypedDict):
    url: str
    method: str
    body: Any
    accepts: list[PaymentRequirement]


class PaymentHandlerResult(TypedDict):
    headers: dict[str, str]


SyncPaymentHandler = Callable[[PaymentHandlerContext], PaymentHandlerResult]
AsyncPaymentHandler = Callable[[PaymentHandlerContext], Awaitable[PaymentHandlerResult]]
PaymentHandler = Union[SyncPaymentHandler, AsyncPaymentHandler]


def parse_payment_required(headers: Mapping[str, str], body: Any) -> dict[str, Any]:
    encoded = headers.get("PAYMENT-REQUIRED") or headers.get("payment-required")
    if encoded:
        try:
            value = json.loads(base64.b64decode(encoded).decode("utf-8"))
        except Exception as exc:
            raise ValueError("Invalid PAYMENT-REQUIRED header") from exc
        if isinstance(value, dict):
            return value
        raise ValueError("Invalid PAYMENT-REQUIRED header")
    if isinstance(body, dict):
        return body
    if isinstance(body, (bytes, bytearray)):
        body = body.decode("utf-8", errors="replace")
    if isinstance(body, str):
        try:
            value = json.loads(body)
        except Exception:
            return {}
        if isinstance(value, dict):
            return value
    return {}
